security/advancedtls/advancedtls_test.go

```diff
@@ -46,14 +47,249 @@ func Test(t *testing.T) {
+	cs := &certStore{}
+	err := cs.loadCerts()
+	if fake.wantError {
+		return nil, fmt.Errorf("bad fakeIdentityProvider")
+	}
+	if err != nil {
+		return nil, fmt.Errorf("failed to load certs: %v", err)
+	}
+	if fake.isClient {
+		if fake.wantMultiCert {
+			return &certprovider.KeyMaterial{Certs: []tls.Certificate{cs.clientPeer1, cs.clientPeer2}}, nil
+		}
+		return &certprovider.KeyMaterial{Certs: []tls.Certificate{cs.clientPeer1}}, nil
+	}
+	if fake.wantMultiCert {
+		return &certprovider.KeyMaterial{Certs: []tls.Certificate{cs.serverPeer1, cs.serverPeer2}}, nil
+	}
+	return &certprovider.KeyMaterial{Certs: []tls.Certificate{cs.serverPeer1}}, nil
+}
+
+func (fake fakeIdentityProvider) Close() {}
+
+func (s) TestClientOptionsConfig(t *testing.T) {
+	tests := []struct {
+		desc        string
+		clientVType VerificationType
+		IdentityCertificateOptions
+		RootCertificateOptions
+		wantError bool
+	}{
+		{
+			desc:        "Skip default verification and provide no root credentials",
+			clientVType: SkipVerification,
+			wantError:   true,
+		},
+		{
+			desc:        "More than one fields in RootCertificateOptions is specified",
+			clientVType: CertVerification,
+			RootCertificateOptions: RootCertificateOptions{
+				RootCACerts:  x509.NewCertPool(),
+				RootProvider: fakeRootProvider{},
+			},
+			wantError: true,
+		},
+		{
+			desc:        "More than one fields in IdentityCertificateOptions is specified",
+			clientVType: CertVerification,
+			IdentityCertificateOptions: IdentityCertificateOptions{
+				GetIdentityCertificatesForClient: func(*tls.CertificateRequestInfo) (*tls.Certificate, error) {
+					return nil, nil
+				},
+				IdentityProvider: fakeIdentityProvider{},
+			},
+			wantError: true,
+		},
+		{
+			desc: "Specify GetIdentityCertificatesForServer",
+			IdentityCertificateOptions: IdentityCertificateOptions{
+				GetIdentityCertificatesForServer: func(*tls.ClientHelloInfo) ([]*tls.Certificate, error) {
+					return nil, nil
+				},
+			},
+			wantError: true,
+		},
+		{
+			desc:        "Use system default if no fields in RootCertificateOptions is specified",
+			clientVType: CertVerification,
+		},
+		{
+			desc:        "Good case with mutual TLS",
+			clientVType: CertVerification,
+			RootCertificateOptions: RootCertificateOptions{
+				RootProvider: fakeRootProvider{},
+			},
+			IdentityCertificateOptions: IdentityCertificateOptions{
+				IdentityProvider: fakeIdentityProvider{},
+			},
+		},
+	}
+	for _, test := range tests {
+		test := test
+		t.Run(test.desc, func(t *testing.T) {
+			clientOptions := &ClientOptions{
+				VType:           test.clientVType,
+				IdentityOptions: test.IdentityCertificateOptions,
+				RootOptions:     test.RootCertificateOptions,
+			}
+			clientConfig, err := clientOptions.config()
+			if (err != nil) != test.wantError {
+				t.Fatalf("Error mismatch in calling clientOptions.config, wantError =%v, got %v", test.wantError, err != nil)
```

Comment:
Please identify the function which failed and its inputs in error messages.
```
t.Fatalf("ClientOptions{%v}.config() = %v, wantErr %v", clientOptions, err, test.wantErr)
```
Same comment applies to other error messages also.